File: src/risk/position_guard.py
import logging

logger = logging.getLogger(__name__)


class PositionGuard:
    def __init__(self, max_base_exposure_pct: float = 80.0,
                 min_usdt_reserve: float = 50.0, total_capital: float = 200.0):
        self.max_base_exposure_pct = max_base_exposure_pct
        self.min_usdt_reserve = min_usdt_reserve
        self.total_capital = total_capital

    def base_exposure_pct(self, current_base: float, base_price: float,
                          equity: float = 0.0) -> float:
        base_value = current_base * base_price
        base = equity if equity > 0 else self.total_capital
        return (base_value / base) * 100

    def can_place_order(self, current_base: float, base_price: float,
                        current_usdt: float, order_usdt: float,
                        equity: float = 0.0) -> bool:
        # Reject negative or zero order amounts
        if order_usdt <= 0:
            logger.warning(f"Rejected order with non-positive amount: {order_usdt}")
            return False

        if (current_usdt - order_usdt) < self.min_usdt_reserve:
            return False
        base = equity if equity > 0 else self.total_capital
        new_base_value = (current_base * base_price) + order_usdt
        new_exposure_pct = (new_base_value / base) * 100
        if new_exposure_pct > self.max_base_exposure_pct:
            return False
        return True

File: trading-engine-core/src/risk/position_guard.rs
pub struct PositionGuard {
    max_exposure_pct: f64,
    min_usdt_reserve: f64,
    total_capital: f64,
}

impl PositionGuard {
    pub fn new(max_exposure_pct: f64, min_usdt_reserve: f64, total_capital: f64) -> Self {
        Self { max_exposure_pct, min_usdt_reserve, total_capital }
    }

    pub fn can_place_order(
        &self,
        current_base_value: f64,
        base_price: f64,
        current_usdt: f64,
        order_usdt: f64,
        equity: f64,
    ) -> bool {
        let equity = if equity > 0.0 { equity } else { self.total_capital };

        if current_usdt - order_usdt < self.min_usdt_reserve {
            return false;
        }

        let new_base_value = current_base_value + order_usdt / base_price;
        let exposure_pct = (new_base_value * base_price) / equity * 100.0;
        if exposure_pct > self.max_exposure_pct {
            return false;
        }

        true
    }
}
